### PR TITLE
CS 286 confirm your training provider get name from roatp

### PR DESCRIPTION
**Pull request by Pauljgraham**

### CODE REVIEW

File: src/SFA.DAS.ApprenticeCommitments.UnitTests/Application/Services/CommitmentsV2ServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        [Test, AutoData]
+        public async Task WhenApprenticeshipIsNotFound(
+            long accountId,
+            long apprenticeshipId,
+            [Frozen] Mock<IInternalApiClient<CommitmentsV2Configuration>> client)
+        {
+            var sut = new CommitmentsV2Service(client.Object);
+            Func<Task> func = async () => await sut.GetApprenticeshipDetails(accountId, apprenticeshipId);
+            
+            func.Should().Throw<HttpRequestContentException>();
```
Comment (benagain) [inline src/SFA.DAS.ApprenticeCommitments.UnitTests/Application/Services/CommitmentsV2ServiceTests.cs:26]: There is an alternative syntax which may (or may not!) be appealing

```suggestion
            sut.Invoking(s => s.GetApprenticeshipDetails(accountId, apprenticeshipId))
                .Should().Throw<HttpRequestContentException>();

```

File: src/SFA.DAS.ApprenticeCommitments.UnitTests/Application/Services/TrainingProviderServiceTests.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            
+            Func<Task> func = async () => await sut.GetTrainingProviderDetails(trainingProviderId);
+
+            func.Should().Throw<HttpRequestContentException>();
+        }
+
+        [Test, AutoData]
+        public async Task WhenASingleTrainingProvidersIsFound(
+            long trainingProviderId,
+            TrainingProviderResponse result,
```
Comment (benagain) [inline src/SFA.DAS.ApprenticeCommitments.UnitTests/Application/Services/TrainingProviderServiceTests.cs:48]: I like these tests, very clear what they are doing 💯

File: src/SFA.DAS.ApprenticeCommitments/Application/Commands/CreateApprenticeship/CreateApprenticeshipCommandHandler.cs [hunk 3]
```diff
@@ -5,44 +5,45 @@
...
             _apprenticeLoginService = apprenticeLoginService;
             _commitmentsService = commitmentsV2Service;
+            _trainingProviderService = trainingProviderService;
         }
 
         public async Task<Unit> Handle(
             CreateApprenticeshipCommand command,
             CancellationToken cancellationToken)
         {
+            (TrainingProviderResponse trainingProvider, Apis.CommitmentsV2InnerApi.ApprenticeshipResponse apprentice) = await GetExternalData(command);
```
Comment (benagain) [inline src/SFA.DAS.ApprenticeCommitments/Application/Commands/CreateApprenticeship/CreateApprenticeshipCommandHandler.cs:35]: It's possible to use `var` here too:

```suggestion
            var (trainingProvider, apprentice) = await GetExternalData(command);
```

File: src/SFA.DAS.ApprenticeCommitments/Application/Commands/CreateApprenticeship/CreateApprenticeshipCommandHandler.cs [hunk 4]
```diff
@@ -57,5 +58,18 @@ public class CreateApprenticeshipCommandHandler : IRequestHandler<CreateApprenti
...
+
+        private async Task<(TrainingProviderResponse, Apis.CommitmentsV2InnerApi.ApprenticeshipResponse)> GetExternalData(CreateApprenticeshipCommand command)
+        {
+            var trainingProviderTask = _trainingProviderService.GetTrainingProviderDetails(command.TrainingProviderId);
+
+            var apprenticeTask = _commitmentsService.GetApprenticeshipDetails(
+                command.EmployerAccountId,
+                command.ApprenticeshipId);
+
+            await Task.WhenAll(trainingProviderTask, apprenticeTask);
```
Comment (benagain) [inline src/SFA.DAS.ApprenticeCommitments/Application/Commands/CreateApprenticeship/CreateApprenticeshipCommandHandler.cs:70]: Good use of concurrency 👍

**Review by benagain — APPROVED** [review]
Good PR 👍

File: src/SFA.DAS.ApprenticeCommitments.Api.AcceptanceTests/Steps/AddApprenticeshipSteps.cs [hunk 5]
```diff
@@ -113,7 +139,18 @@ public void ThenTheRequestToTheInnerApiWasMappedCorrectly()
...
+        [Then(@"the Training Provider Name should be '(.*)'")]
+        public void ThenTheTrainingProviderNameShouldBe(string trainingProviderName)
+        {
+            var logs = _context.InnerApi.MockServer.LogEntries;
+            logs.Should().HaveCount(1);
+
+            var innerApiRequest = JsonConvert.DeserializeObject<CreateApprenticeshipRequestData>(
+                logs.First().RequestMessage.Body);
+
+            innerApiRequest.TrainingProviderName.Should().Be(trainingProviderName);
```
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/SFA.DAS.ApprenticeCommitments.Api.AcceptanceTests/Steps/AddApprenticeshipSteps.cs:153]: Good point but should be in a seperate step. I will check

[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/SFA.DAS.ApprenticeCommitments.Api.AcceptanceTests/Steps/AddApprenticeshipSteps.cs:153]: wasn't there but I added it now

File: src/SFA.DAS.ApprenticeCommitments.UnitTests/Application/Services/TrainingProviderServiceTests.cs (hunk 2 above)
Comment (Pauljgraham) [reply src/SFA.DAS.ApprenticeCommitments.UnitTests/Application/Services/TrainingProviderServiceTests.cs:48]: I hate the syntax of teh Setup. Prefer to use write a Fixture, but these aren't liked. The only compromise I find it to put the setup code in private functions

File: src/SFA.DAS.ApprenticeCommitments.UnitTests/Application/Services/CommitmentsV2ServiceTests.cs (hunk 1 above)
Comment (Pauljgraham) [reply src/SFA.DAS.ApprenticeCommitments.UnitTests/Application/Services/CommitmentsV2ServiceTests.cs:26]: I prefer your suggestion, easier to read and makes the intent clearer

File: src/SFA.DAS.ApprenticeCommitments/Application/Commands/CreateApprenticeship/CreateApprenticeshipCommandHandler.cs (hunk 3 above)
Comment (Pauljgraham) [reply src/SFA.DAS.ApprenticeCommitments/Application/Commands/CreateApprenticeship/CreateApprenticeshipCommandHandler.cs:35]: thx, great suggestion

**Review by benagain — APPROVED** [review]

**Review by benagain — APPROVED** [review]

**Pull request merged**